tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc

```diff
@@ -0,0 +1,834 @@
+      per_device_collector_[i].Export(start_walltime_ns_, start_gputime_ns_,
+                                      end_gputime_ns, &device_plane,
+                                      &host_plane);
+      GetDeviceCapabilities(i, &device_plane);
+      NormalizeTimeStamps(&device_plane, start_walltime_ns_);
+    }
+    NormalizeTimeStamps(&host_plane, start_walltime_ns_);
+  }
+
+ private:
+  std::atomic<int> num_callback_events_;
+  std::atomic<int> num_activity_events_;
+  uint64 start_walltime_ns_;
+  uint64 start_gputime_ns_;
+
+  mutex aggregated_events_mutex_;
+  absl::flat_hash_map<uint32, RocmTracerEvent> aggregated_events_
+      TF_GUARDED_BY(aggregated_events_mutex_);
+
+  // We need to create a map of
+  //  event.device_id -> index into per_device_collector_ array
+  // The event.device_id returned by the RocmTracer is the physical
+  // device_id and not the logical device_id. Say for example we are
+  // running on a node with 8 GPUs. The expected physical device_id(s)
+  // for those 8 GPUs would be 0,1,2,3,4,5,6,7. On such a node, if we
+  // run a test with HIP_VISIBLE_DEVICES=5, then "options.num_gpus_ == 1",
+  // but the event.device_id field will have 5 in it!
+  // So the event.device_id can be thought of as the physical device id
+  // and the index can be thought of as the logical device id.
+  // We cannot determine the actual phsyical device id logical device id
+  // mapping here, so we determine it empirically
+  std::map<uint32, uint32> device_id_map_;
+  uint32 next_logical_device_id_;
+
+  bool IsEventTypeWithoutHCCActivityRecordCallback(RocmTracerEventType type) {
+    switch (type) {
+      case RocmTracerEventType::MemoryAlloc:
+        return true;
+        break;
+      default:
+        break;
+    }
+    return false;
+  }
+
+  struct PerDeviceCollector {
+    void AddEvent(const RocmTracerEvent& event) {
+      mutex_lock lock(events_mutex);
+      events.emplace_back(event);
+    }
+
+    void SortByStartTime() {
+      mutex_lock lock(events_mutex);
+      std::sort(
+          events.begin(), events.end(),
+          [](const RocmTracerEvent& event1, const RocmTracerEvent& event2) {
+            return event1.start_time_ns < event2.start_time_ns;
+          });
+    }
+
+    void Export(int32 device_ordinal, uint64 start_walltime_ns,
+                uint64 start_gputime_ns, StepStats* step_stats) {
+      mutex_lock lock(events_mutex);
+      absl::flat_hash_map<std::pair<uint64 /*stream_id*/, RocmTracerEventType>,
+                          DeviceStepStats*>
+          per_stream_dev_stats;
+
+      DeviceStepStats* generic_stream_dev_stats = nullptr;
+      DeviceStepStats* all_streams_dev_stats = nullptr;
+      DeviceStepStats* memcpy_dev_stats = nullptr;
+      DeviceStepStats* sync_dev_stats = nullptr;
+
+      for (const RocmTracerEvent& event : events) {
+        DumpRocmTracerEvent(event, start_walltime_ns, start_gputime_ns);
+
+        NodeExecStats* ns = new NodeExecStats;
```

Comment:
Prefer to use a std::unique_ptr, so that this will not leak if any other devs changes the (long) code sequence below to return early.

Finally, release() the pointer to swap().